File: graph/S1_2667.py



N = int(input())

def mapping(graph,visited, x, y, cnt):
    # print(x, y)
    visited[x][y] = 1
    cnt += 1
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if nx < 0 or nx >= N or ny < 0 or ny >= N:
            continue
        if visited[nx][ny] == 1:
            continue
        if graph[nx][ny] == 0:
            continue
        cnt = mapping(graph,visited, nx, ny, cnt)
    return cnt



graph = []
visited = [[0]*N for _ in range(N)]

for i in range(N):
    graph.append(list(map(int, input())))

# print(graph)

dx = [1, -1, 0, 0]
dy = [0, 0, -1, 1]

arr = []
home = 0
for i in range(N):
    for j in range(N):
        if graph[i][j] == 1 and visited[i][j] == 0:
            cnt = 0
            cnt = mapping(graph,visited, i, j, cnt)
            arr.append(cnt)
            home += 1
print(home)
arr.sort()
for i in range(len(arr)):
    print(arr[i])

File: graph/S1_2667.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'
int dx[] = {1, -1, 0, 0};
int dy[] = {0, 0, 1, -1};
int depth_first_search(vector<vector<bool>> &graph, vector<vector<bool>> &visited, int x, int y, int n, int cnt)
{
    if (visited[x][y])
        return 0;
    int ret = 0;
    visited[x][y] = true;
    for (int i = 0 ; i < 4 ; ++i)
    {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (nx < 0 || nx >= n || ny < 0 || ny >= n)
            continue;
        if (graph[nx][ny])
            ret += depth_first_search(graph, visited, nx, ny, n, cnt + 1);
    }
    return 1 + ret;
}

int main()
{
    int n;
    cin >> n;
    vector<vector<bool>> graph(n, vector<bool>(n, false)), visited(n, vector<bool>(n, false));

    for (int i = 0 ; i < n ; ++i)
    {
        string str;
        cin >> str;
        for (int j = 0 ; j < n ; ++j)
        {
            str[j] == '1' ? graph[i][j] = true : graph[i][j] = false;
        }
    }
    vector<int> housesNum;
    int count = 0;
    for (int i = 0 ; i < n ; ++i)
    {
        for (int j = 0 ; j < n ; ++j)
        {   
            if (!visited[i][j] && graph[i][j]){
                count++;
                int num = depth_first_search(graph, visited, i, j, n, 0);
                housesNum.push_back(num);
                // memset(visited, false, sizeof(visited));
            }
        }
    }
    cout << count << endl;
    sort(housesNum.begin(), housesNum.end());
    for (auto i : housesNum)
        cout << i << endl;
    
}